src/hotspot/share/runtime/mutex.cpp

```diff
@@ -372,17 +371,56 @@ Mutex* Mutex::get_least_ranked_lock_besides_this(Mutex* locks) {
       res = tmp;
     }
   }
+  assert(res != this, "invariant");
+  return res;
+}
+
+// Tests for rank violations that might indicate exposure to deadlock.
+void Mutex::check_rank(Thread* thread) {
+  assert(this->rank() >= 0, "bad lock rank");
+  Mutex* locks_owned = thread->owned_locks();
+
   if (!SafepointSynchronize::is_at_safepoint()) {
-    // In this case, we expect the held locks to be
-    // in increasing rank order (modulo any native ranks)
-    for (tmp = locks; tmp != NULL; tmp = tmp->next()) {
+    // We expect the locks already acquired to be in increasing rank order,
+    // modulo locks of native rank or acquired in try_lock_without_rank_check()
+    for (Mutex* tmp = locks_owned; tmp != NULL; tmp = tmp->next()) {
       if (tmp->next() != NULL) {
-        assert(tmp->rank() == Mutex::native ||
-               tmp->rank() <= tmp->next()->rank(), "mutex rank anomaly?");
+        assert(tmp->rank() == Mutex::native || tmp->rank() < tmp->next()->rank()
+               || tmp->skip_rank_check(), "mutex rank anomaly?");
       }
     }
   }
-  return res;
+
+  // Locks with rank native or suspend_resume are an exception and are not
+  // subject to the verification rules.
+  bool check_can_be_skipped = this->rank() == Mutex::native || this->rank() == Mutex::suspend_resume
+                              || SafepointSynchronize::is_at_safepoint();
+  if (owned_by_self()) {
+    // wait() case
+    Mutex* least = get_least_ranked_lock_besides_this(locks_owned);
```

Comment:
Wouldn't you want to test check_can_be_skipped and skip all of this?

Reply:
For the wait() case, today we always check that the thread doesn't wait while holding a Mutex/Monitor of rank less than or equal to special, and I didn't want to change that.
But I could do it for the "lock()/lock_without_safepoint_check()/try_lock()" case, and return before calling get_least_ranked_lock().

Reply:
too many conditions!  Now I see it.  Okay.